src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/KeyVariantByVariantShard.java

```diff
@@ -0,0 +1,42 @@
+package org.broadinstitute.hellbender.engine.dataflow.transforms;
+
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.KV;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import org.broadinstitute.hellbender.engine.dataflow.datasources.VariantShard;
+import org.broadinstitute.hellbender.utils.variant.Variant;
+
+import java.util.List;
+
+/**
+ * KeyVariantByVariantShard takes a PCollection of variants and keys each read by the variants shards it overlaps.
+ * Note that if a variant span several shards multiple KVs will be produced. This part of the larger transform to join
+ * reads with overlapping variants.
+ *
+ * |---- shard 0 -----|---- shard 1 -----|---- shard 2 -----|---- shard 3 -----|---- shard 4 -----|
+ *           |-------------- variant 1 --------------|
+ *  results in
+ *  KV<shard 0, variant 1>
+ *  KV<shard 1, variant 1>
+ *  KV<shard 2, variant 1>
+ */
+public class KeyVariantByVariantShard extends PTransform<PCollection<Variant>, PCollection<KV<VariantShard, Variant>>> {
```

Comment:
How about writing `KeyLocationByOverlappedShard` and using that instead of both `KeyReadByVariantShard` and `KeyVariantByVariantShard`? It can be generic so that the returned collection would still have the expected type (`KV<Location,Variant>`). 

This way we have less code, so less work and fewer bug opportunities.